subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ExecutionStateChanges.java

```diff
@@ -16,18 +16,18 @@
 
 package org.gradle.internal.execution.history.changes;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableListMultimap;
-import org.gradle.internal.change.ChangeVisitor;
 
 /**
  * Represents the complete changes in execution state
  */
 public interface ExecutionStateChanges {
 
     /**
-     * Visits any change to inputs or outputs.
+     * Returns all change messages for inputs and outputs.
      */
-    void visitAllChanges(ChangeVisitor visitor);
+    ImmutableList<String> getAllChangeMessages();
```

Comment:
Why change this?

Reply:
Why not? Returning the list of change messages seems to be simpler than requiring a visitor - especially if we already determined all the change message when resolving the changes.